### PR TITLE
Bugfix 3.4.1:  Clean-up edge cases of Scheduler shutdown

### PR DESCRIPTION
**Pull request by matthewvon**
This PR addresses edge cases relating to shutting down the Scheduler:

- there was the potential for other features to submit tasks to the Scheduler as it was attempting to shutdown.  This could lead to crashes and/or stuck shutdown in other features (note:  no cases of this have been seen recently)

- beginShutdown() and shutdown() were only clearing tasks that were in the working or queued state.  Tasks sitting in the work FIFOs were not addressed.   FIFO tasks could be added to the Scheduler after stop() was called on the io_context and race condition allowed adding after io_context destroyed.  (io_context worker threads that were shutting down first called drain() which attempted to post work from the FIFOs).

### CODE REVIEW

**Comment by matthewvon** [comment]
All balls blue:  http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/2361/

File: arangod/Scheduler/Scheduler.cpp [hunk 1]
```diff
@@ -208,24 +208,45 @@ Scheduler::~Scheduler() {
...
 
     // reduce number of queued now
     decQueued();
 
+    // this post is coming late in application shutdown,
+    //  might be essential ... process in-line after heavy complaining
+    LOG_TOPIC(WARN, Logger::THREADS)
+      << "Scheduler::post() called after isStopping() set.";
+    TRI_ASSERT(false);
+    std::abort();   /// temporary ... debugging within Jenkins
```
Comment (goedderz) [inline arangod/Scheduler/Scheduler.cpp]: @matthewvon You forgot removing the `abort` here.

File: arangod/Scheduler/Scheduler.cpp [hunk 2]
```diff
@@ -525,6 +575,30 @@ void Scheduler::shutdown() {
...
+    _ioContext->post([] {
+        LOG_TOPIC(DEBUG, Logger::THREADS) << "Wakeup alarm";
+        /*wakeup call for scheduler thread*/
+      });
+
+    std::this_thread::yield();
+    // we can be quite generous here with waiting...
+    // as we are in the shutdown already, we do not care if we need to wait
+    // for a bit longer
+    std::this_thread::sleep_for(std::chrono::microseconds(20000));
```
Comment (goedderz) [inline arangod/Scheduler/Scheduler.cpp]: Why not write 20 milliseconds instead of 20000 microseconds?

File: arangod/Scheduler/Scheduler.cpp (hunk 1 above)
Comment (matthewvon) [reply arangod/Scheduler/Scheduler.cpp]: Corrected, thank you.

File: arangod/Scheduler/Scheduler.cpp (hunk 2 above)
Comment (matthewvon) [reply arangod/Scheduler/Scheduler.cpp]: The original author of this file has consistently used microseconds throughout this file.  The lines above a copy/paste of a similar code.  I am going to leave this as microseconds to maintain consistency.

**Comment by matthewvon** [comment]
still blue balls:  http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/2391/

**Comment by matthewvon** [comment]
@jsteemann If we were ready for 3.4.1 changes, would this pass your inspection?  I do not want to start on devel version until this one is in happy shape.

**Pull request closed without merging**